kernel/constr.ml

```diff
@@ -911,8 +916,12 @@ let compare_head_gen_leq_with kind1 kind2 leq_universes leq_sorts eq leq nargs t
   | Ind (c1,u1), Ind (c2,u2) -> Ind.CanOrd.equal c1 c2 && leq_universes (Some (GlobRef.IndRef c1, nargs)) u1 u2
   | Construct (c1,u1), Construct (c2,u2) ->
     Construct.CanOrd.equal c1 c2 && leq_universes (Some (GlobRef.ConstructRef c1, nargs)) u1 u2
-  | Case (_,p1,iv1,c1,bl1), Case (_,p2,iv2,c2,bl2) ->
-    eq 0 p1 p2 && eq_invert (eq 0) (leq_universes None) iv1 iv2 && eq 0 c1 c2 && Array.equal (eq 0) bl1 bl2
+  | Case (ci1,u1,pms1,p1,iv1,c1,bl1), Case (ci2,u2,pms2,p2,iv2,c2,bl2) ->
+    (** FIXME: what are we doing with u1 = u2 ? *)
+    Ind.CanOrd.equal ci1.ci_ind ci2.ci_ind && leq_universes (Some (GlobRef.IndRef ci1.ci_ind, 0)) u1 u2 &&
```

Comment:
That's a good question. We should be restrictive on cases/eliminators and ask for *equality* of universes, not just leq. Consider an inductive with covariant universes like `Record packType@{i} := pack { aType : Type@{i} }`. 
Allowing subtyping would permit `match p in packType@{i} return ... with pack (aType:Type@{i}) => f aType` <= match p in packType@{i'} return ... with => True` while `i < i'` (assuming `f := fun (_ : aType@{i}) => True`). Using the "identity coercion" semantics of cumulative inductives this would mean we can lift the left hand side to the right hand-side: however `match lift_i_i' p in packType@{i'} return ... with pack (aType : Type@{i'}) => f aType end` would be ill-typed. The rule in our consistency proof (https://hal.inria.fr/hal-01615123v3/document, p8) also forces the same universes. Maybe we could relax this for irrelevant universes (after all, that's what template polymorphism implements), but it's not been studied yet, so I'd rather err on the safe side.

Reply:
There is no equality function argument passed to the toplevel definition though... Also, the current form of the code is equivalent to what we were doing before with the eta-expanded return clause. Shouldn't we add this tweak in another PR?

Reply:
Currently, it allows two matches on `packType@{i}` and `packType@{i'}` to be considered equal which I think is wrong. It's not the same as the previous code where we were comparing domains of lambdas using `eq`, I think.

Reply:
Actually, despite the name, in the kernel all calls are performed with universe equality rather than cumulativity. So at least from a TCB point of view, this shouldn't matter. Outside of the kernel, the only user of this function I could find which doesn't use equality is `test_constr_universes`, and even there it's not clear how often we call it with the `leq` flag set to false.

Reply:
Indeed, for example `leq_univs` below is clearly incomplete as it passes `eq_universes`. Let's leave this for another PR, the current code as a whole implements the same behavior.